src/Compiler/Optimize/Optimizer.fs

```diff
@@ -3247,17 +3247,24 @@ and TryDevirtualizeApplication cenv env (f, tyargs, args, m) =
     | Expr.Val (v, _, _), [ty], _ when CanDevirtualizeApplication cenv v g.generic_equality_withc_inner_vref ty args ->
         let tcref, tyargs = StripToNominalTyconRef cenv ty
         match tcref.GeneratedHashAndEqualsWithComparerValues, args with
-        | Some (_, _, withcEqualsVal), [comp; x; y] -> 
+        | Some (_, _, _, Some withcEqualsExactVal), [comp; x; y] -> 
+            // push the comparer to the end
+            let args2 = [x; mkRefTupledNoTypes g m [y; comp]]
+            Some (DevirtualizeApplication cenv env withcEqualsExactVal ty tyargs args2 m)
+        | Some (_, _, withcEqualsVal, _ ), [comp; x; y] -> 
```

Comment:
If I read things correctly, this code path is being taken when `withcEqualsExactVal` is None, i.e. when it was being unpickled from a different assembly.

And the line below actually boxes the argument.


=> Does this mean that the benefits are not used in a cross-assembly scenario?
The "other" assembly might be a `Project.Domain` library with domain objects, but also value types from fslib such as `ValueOption`.


I scanned the `GenericComparison.fs` test definition and did not find a test case working with more than 1 assembly to verify IL in a cross-assembly scenario, but I might have overlooked it.

Reply:
Okay so this also works now thanks to pushing this overload to the ad-hoc table. Tests are here: https://github.com/dotnet/fsharp/pull/16857/files#diff-df081675edf7d8decef09032e5ea509176d734fa6cc175ca888bc6b965acbe98